data/migrations/V0078__add_indexes_on_fec_fitem_sched_b_tables.sql

```diff
@@ -0,0 +1,1381 @@
+CREATE INDEX IF NOT EXISTS idx_sched_b_1999_2000_recipient_name_tex ON disclosure.fec_fitem_sched_b_1999_2000 USING gin (recipient_name_text);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1999_2000_rpt_yr ON disclosure.fec_fitem_sched_b_1999_2000 USING btree (rpt_yr);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1999_2000_line_num ON disclosure.fec_fitem_sched_b_1999_2000 USING btree (line_num COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_1999_2000_disb_dt ON disclosure.fec_fitem_sched_b_1999_2000 USING btree (disb_dt);
+CREATE INDEX IF NOT EXISTS idx_sched_b_1999_2000_coalesce_disb_dt ON disclosure.fec_fitem_sched_b_1999_2000 USING btree ((COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)));
+ANALYZE disclosure.fec_fitem_sched_b_1999_2000;
+-- -------------------------------
+-- -------------------------------
+--   fec_fitem_sched_b_2001_2002
+-- -------------------------------
+-- -------------------------------
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_clean_recipient_cmte_id ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (clean_recipient_cmte_id COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_recipient_st ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (recipient_st COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_recipient_city ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (recipient_city COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_cmte_id ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (cmte_id COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_disb_amt ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (disb_amt);
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_disbursement_description_text ON disclosure.fec_fitem_sched_b_2001_2002 USING gin (disbursement_description_text);
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_image_num ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (image_num COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_recipient_name_text ON disclosure.fec_fitem_sched_b_2001_2002 USING gin (recipient_name_text);
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_rpt_yr ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (rpt_yr);
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_line_num ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (line_num COLLATE pg_catalog."default");
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_disb_dt ON disclosure.fec_fitem_sched_b_2001_2002 USING btree (disb_dt);
+CREATE INDEX IF NOT EXISTS idx_sched_b_2001_2002_coalesce_disb_dt ON disclosure.fec_fitem_sched_b_2001_2002 USING btree ((COALESCE(disb_dt, '9999-12-31'::date::timestamp without time zone)));
+ANALYZE disclosure.fec_fitem_sched_b_2001_2002;
+
+
+

```

Comment:
I'm curious about the change in style of how the indexes are created once you hit 2003; was there a reason for not just using the simpler `CREATE INDEX` statements as above?

Another thing about the index creation - these don't quite mirror what we have on the current `ofec_sched_b_xxxxx` tables, where the indexes are multicolumn to efficiently support the paging, filtering, and sorting we have on the site.  I know the PostgreSQL documentation says that PostgreSQL is usually able to combine indexes effectively on its own, but our research into this has shown that's not been the case with the site and we've needed to create the indexes with each combination of columns.  I'm happy to review this further with you if it'd be helpful to walk through and demonstrate what this means if anything is unclear!

I call this out because I'm concerned that we'll lose some of the performance gains we've made, especially if we're still looking to enable multi-cycle filtering/searching in the future.

Reply:
The change in style of how the indexes are created after year 2003 (Also stated in the migration file itself) is as followed:
Creating all these indexes during release time will take LONG time and also cause performance problems, especially for the bigger partition tables.  Since tables are not being used before switch, indexes had been built during non-peak hour.  However, migration file is still submitted to keep the track of the integrity of the migration process.  Suppress a particular error reporting (error: 42P07 creating indexes if they already exist without error out).   Using create index if not exists will also generate a LOT of warning message.  Indexes for the smaller tables will be created normally using the migration file itself.

Reply:
For the sample DB, this is another issue Joseph Kuan is currently working on.  He is planning on using a local postgresql database, which we had kept all the base tables data current to server as the sample DB instead of using the sample DB script to load small set of sample data.

Reply:
Helen and Jean had conducted analysis on the multiple column indexes vs combined single column indexes queries.  Postgresql DB had combine indexes effectively on its own with acceptable performance.  So we decided to use the single column indexes and let the DB engine handle the combination.  However, I did not consider the pagination/sorting enough during the analysis.  I am in the process of re-evaluating the indexes now.

Reply:
D'oh, I completely missed the note at the top of the migration file regarding the index creation, I am sorry about that, @fecjjeng!  Thanks for bringing it to my attention. :-)

I'm curious to hear more about what's happening with the sample DB, but if that work is happening outside of this we can follow up where it's more appropriate instead of focusing on it here.

If you need any help with the index analysis please let me know!

Reply:
@fecjjeng yes, the `COALESCE`-based indexes are definitely needed against the `disb_dt`, at least as the API endpoint is structured right now.  I can see not wanting to include the older style, which I think is okay for now, but please be mindful if the endpoint changes in the future.  In other words, if the query construction changes and `disb_dt` ends up being used directly again, the regular style indexes will be needed again.

For the third one, this is because we allow sorting by disbursement amount (labeled as `disb_amt`) in addition to the disbursement date, so we need a set of indexes that make use of it as well.  `last_disbursement_amount` is in reference to an [additional argument and field](https://github.com/fecgov/openFEC/blob/develop/webservices/args.py#L511) added to the API to help with the sorting and pagination.  It is used to keep track of the last amount of the previous page when paging through results sorted by disbursement amount (and there's a corresponding one for `last_disbursement_date` as well).  This is because of the way the paging is ultimately handled when the query is constructed.

Most of the paging code itself happens with the [`SeekCoalescePaginator`](https://github.com/fecgov/openFEC/blob/develop/webservices/utils.py#L83) in this case, which is accessed by the [`ItemizedResource`](https://github.com/fecgov/openFEC/blob/develop/webservices/common/views.py#L56) for resources that are specific to the itemized schedule, as in this case.

All of this is to say that yes, a set of indexes is needed with `disb_amt`. :-)

Did I answer your questions?  Please let me know if anything is still unclear!

Reply:
Also, don't forget about the documentation around how the expression based indexes are supported, that can help explain things too!  [Enabling an expression based sort](https://github.com/fecgov/openFEC/wiki/Enabling-an-expression-based-sort)

Reply:
Yes, I remembered we were working on the function-based indexes for disb_dt.
I realized that my first test does not show disb_amt in the query since I was calling the API in a mixed way.  Now when I select disbursement_amount in the sort field and run the next page with last_disbusement_amount, the query does include the disb_amt in the where AND order by field.  Then I noticed a new question, the following is the where/order by clause I got:
WHERE disclosure.fec_fitem_sched_b.two_year_transaction_period = %(two_year_transaction_period_1)s AND disclosure.fec_fitem_sched_b.cmte_id IN (%(cmte_id_1)s) AND (coalesce(disclosure.fec_fitem_sched_b.disb_amt, NULL), disclosure.fec_fitem_sched_b.sub_id) < (%(param_1)s, %(param_2)s) ORDER BY disclosure.fec_fitem_sched_b.disb_amt DESC, disclosure.fec_fitem_sched_b.sub_id DESC
LIMIT %(param_3)s

The where clause has coalesce for disb_amt but the order by clause does not.  I remember last time when these two clause does not match, the performance is not optimized for disb_dt.  Should we made the indexes to use the coalesce disb_amt and change the models/itemized.py to add the coalesce disb_amt to make it consistene?

Reply:
Oh good question, @fecjjeng!  We should dig into that further to see why a `COALESCE` statement is being generated for use with disbursement amount, especially just to default to `NULL` anyway, that doesn't make sense and causes the very issue we were trying to avoid with using `COALESCE` for the disbursement date.

From the app code side, we're only defining the disbursement date for use with a `COALESCE` in the `ORDER BY` clause ([as seen here](https://github.com/fecgov/openFEC/blob/develop/webservices/common/models/itemized.py#L343-L355)), so this is with the SQL construction side.  Like I said, we should investigate why because that doesn't seem right or useful, but there might be something I'm also missing/forgetting.

Reply:
@ccostino I agree.  I will create a ticket to investigate this issue.  For the indexes, I will create set #1 and Set #3 for now.

Reply:
That sounds great, thanks @fecjjeng! :-)

Reply:
Thanks for all the research @fecjjeng @ccostino 
From Jean's post above, it looks like the COALESCE with disbursement_amount was created when pagination happens.  We definitely should investigate it.